### PR TITLE
AP_RangeFinder: Add driver for NoopLoop rangefinder

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This adds UART driver for Nooploop TOF sense P rangefinder: https://ftp.nooploop.com/software/products/tof/doc/TOFSense_User_Manual_V2.5_zh.pdf

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "AP_RangeFinder.h"
+
+#ifndef AP_RANGEFINDER_NOOPLOOP_ENABLED
+#define AP_RANGEFINDER_NOOPLOOP_ENABLED AP_RANGEFINDER_ENABLED
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.h:6]:
```suggestion
#define AP_RANGEFINDER_NOOPLOOP_ENABLED AP_RANGEFINDER_BACKEND_DEFAULT_ENABLED && BOARD_FLASH_SIZE > 1024
```

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+// distance returned in reading_m
+bool AP_RangeFinder_NoopLoop::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp:55]: This is not what this method returns.

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    while (nbytes-- > 0) {
+        int16_t r = uart->read();
+        if (r < 0) {
+            continue;
+        }
+        uint8_t c = (uint8_t)r;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp:61]:
```suggestion
        uint8_t c;
        if (!uart->read(c)) {
            break;
        }
```

**Review by peterbarker — DISMISSED** [review]
This needs a simulator.  Every othr serial rangefinder has one, and they're usually only a dozen lines.  `SIM_RF_Benewake` looks like a good one to swipe from.  Unsurprisingly, since this driver seems to be swiped from `AP_RangeFinder_Benewake` :-)

**Comment by rishabsingh3003** [comment]
@peterbarker thanks for the review! I have swiped Benewake's SIM :)

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+// distance returned in reading_m
+bool AP_RangeFinder_NoopLoop::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    // read any available lines from the lidar
+    uint32_t nbytes = uart->available();
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp:55]: maybe we should add a limit of 1000 bytes.

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            if (linebuf_len == NOOPLOOP_FRAME_LENGTH) {
+                // calculate checksum
+                uint8_t checksum = 0;
+                for (uint8_t i=0; i<NOOPLOOP_FRAME_LENGTH-1; i++) {
+                    checksum += linebuf[i];
+                }
+                // if checksum matches extract contents
+                if (checksum == linebuf[NOOPLOOP_FRAME_LENGTH-1]) {
+                    // calculate distance
+                    const int32_t dist = (int32_t)(linebuf[8] << 8 | linebuf[9] << 16 | linebuf[10] << 24) / 256;
```
Comment (andyp1per) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp:87]: Probably needs a comment

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
+
+#include <ctype.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define NOOPLOOP_FRAME_HEADER 0x57
+#define NOOPLOOP_FRAME_HEADER_1 0x00
+#define NOOPLOOP_FRAME_LENGTH 16
```
Comment (andyp1per) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.cpp:29]: This should be in the header

File: libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+private:
+    using AP_RangeFinder_Backend_Serial::AP_RangeFinder_Backend_Serial;
+
+
+    // get a reading
+    // distance returned in reading_m
+    bool get_reading(float &reading_m) override;
+
+    uint8_t linebuf[16];
```
Comment (andyp1per) [inline libraries/AP_RangeFinder/AP_RangeFinder_NoopLoop.h:37]: Use NOOPLOOP_FRAME_LENGTH

**Review by rmackay9 — APPROVED** [review]
A few little issues but Rishabh will handle these in a follow-up PR.

**Comment by Hwurzburg** [comment]
@rishabsingh3003 please give me the links to each TOF product that this driver applies to for the wiki...thanks

**Comment by Hwurzburg** [comment]
@rishabsingh3003 bump...see above

**Pull request merged**